src/preallocate.jl

```diff
@@ -0,0 +1,52 @@
+export preallocate
+
+struct PreallocatedMethod{F, Args <: Tuple, N, R}
+    f::F
+    ctx::NTuple{N, R}
+
+    function PreallocatedMethod{F, Args}(f::F, ctx::NTuple{N, R}) where {F, Args, N, R}
+        new{F, Args, N, R}(f, ctx)
+    end
+end
+
+function PreallocatedMethod(f::F, xs...) where F
+    x, record = record_allocations(f, xs...)
+    record = freeze(record)
+    records = (ntuple(_->copy(record), Threads.nthreads() - 1)..., record)
+    ctx = new_replay_ctx.(records)
+    return PreallocatedMethod{F, typeof(xs)}(f, ctx)
+end
+
+function (f::PreallocatedMethod)(xs...)
+    ctx = f.ctx[Threads.threadid()]
+    # RL: Why this is not type stable at all?
```

Comment:
Oh I bet because: `ReplayCtx` has a `Ref{Int}` (which is an `AbstractType`) rather than `Base.RefValue{Int}`

Reply:
yup, should be type stable now...